pulsar-transaction/coordinator/src/test/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriterTest.java

```diff
@@ -613,9 +610,371 @@ public ByteBuf serialize(ArrayList<Integer> dataArray) {
+        SumStrDataSerializer dataSerializer = new SumStrDataSerializer();
+        int batchedWriteMaxSize = 16;
+        int writeCount = 100;
+        int expectedBatchFlushCount = writeCount / (batchedWriteMaxSize / dataSerializer.getSizePerData());
+        int expectedTotalBytesSize = expectedBatchFlushCount * batchedWriteMaxSize;
+        var callbackWithCounter = createCallBackWithCounter();
+        // Create TxnLogBufferedWriter.
+        var txnLogBufferedWriterContext = createTxnBufferedWriterContextWithMetrics(
+                dataSerializer, Integer.MAX_VALUE, batchedWriteMaxSize, Integer.MAX_VALUE);
+        var txnLogBufferedWriter = txnLogBufferedWriterContext.txnLogBufferedWriter;
+        // Add some data.
+        for (int i = 0; i < writeCount; i++){
+            txnLogBufferedWriter.asyncAddData(1, callbackWithCounter.callback, "");
+        }
+        // Wait for all write finish.
+        Awaitility.await().atMost(1, TimeUnit.SECONDS).until(
+                () -> callbackWithCounter.finishCounter.get() + callbackWithCounter.failureCounter.get() == writeCount
+        );
+        int actualBatchFlushCount = txnLogBufferedWriterContext.mockedManagedLedger.writeCounter.get();
+        assertEquals(callbackWithCounter.failureCounter.get(), 0);
+        assertEquals(expectedBatchFlushCount, actualBatchFlushCount);
+        verifyTheCounterMetrics(0, expectedBatchFlushCount,0,0);
+        verifyTheHistogramMetrics(expectedBatchFlushCount, writeCount, expectedTotalBytesSize);
+        // cleanup.
+        releaseTxnLogBufferedWriterContext(txnLogBufferedWriterContext);
+        // after close, verify the metrics change to 0.
+        verifyTheCounterMetrics(0,0,0,0);
+        verifyTheHistogramMetrics(0,0,0);
+    }
+
+    @Test
+    public void testMetricsStatsThatTriggeredByMaxDelayTime() throws Exception {
+        SumStrDataSerializer dataSerializer = new SumStrDataSerializer();
+        int writeCount = 100;
+        int expectedTotalBytesSize = writeCount * dataSerializer.getSizePerData();
+        var callbackWithCounter = createCallBackWithCounter();
+        // Create TxnLogBufferedWriter.
+        var txnLogBufferedWriterContext =
+                createTxnBufferedWriterContextWithMetrics(dataSerializer, Integer.MAX_VALUE,
+                        Integer.MAX_VALUE, 1);
+        var txnLogBufferedWriter = txnLogBufferedWriterContext.txnLogBufferedWriter;
+        // Add some data.
+        for (int i = 0; i < writeCount; i++){
+            txnLogBufferedWriter.asyncAddData(1, callbackWithCounter.callback, "");
+            Thread.sleep(1);
+        }
+        // Wait for all write finish.
+        Awaitility.await().atMost(5, TimeUnit.SECONDS).until(
+                () -> callbackWithCounter.finishCounter.get() + callbackWithCounter.failureCounter.get() == writeCount
+        );
+        int actualBatchFlushCount = txnLogBufferedWriterContext.mockedManagedLedger.writeCounter.get();
```

Comment:
Where are you checking `actualBatchFlushCount == 1` ?

Reply:
> Where are you checking actualBatchFlushCount == 1 ?

Since there has `sleep` at the front, so will flush many times, maybe 40-60 times. and we can only use `mockedManagedLedger.writeCounter.get()` to get the actual count of flush events